presto-spark-base/src/main/java/com/facebook/presto/spark/execution/AbstractPrestoSparkQueryExecution.java

```diff
@@ -0,0 +1,677 @@
+            }
+            else if (executionException instanceof TimeoutException) {
+                failureInfo = Optional.of(toFailure(new PrestoException(EXCEEDED_TIME_LIMIT, "Query exceeded maximum time limit of " + queryTimeout, executionException)));
+            }
+
+            if (!failureInfo.isPresent()) {
+                failureInfo = Optional.of(toFailure(executionException));
+            }
+
+            queryStateTimer.endQuery();
+
+            try {
+                queryCompletedEvent(failureInfo, OptionalLong.empty());
+            }
+            catch (RuntimeException eventFailure) {
+                log.error(eventFailure, "Error publishing query completed event");
+            }
+
+            throw toPrestoSparkFailure(session, failureInfo.get());
+        }
+
+        processShuffleStats();
+
+        ConnectorSession connectorSession = session.toConnectorSession();
+        List<Type> types = getOutputTypes();
+        ImmutableList.Builder<List<Object>> result = ImmutableList.builder();
+        for (Tuple2<MutablePartitionId, PrestoSparkSerializedPage> tuple : rddResults) {
+            Page page = pagesSerde.deserialize(toSerializedPage(tuple._2));
+            checkArgument(page.getChannelCount() == types.size(), "expected %s channels, got %s", types.size(), page.getChannelCount());
+            for (int position = 0; position < page.getPositionCount(); position++) {
+                List<Object> columns = new ArrayList<>();
+                for (int channel = 0; channel < page.getChannelCount(); channel++) {
+                    columns.add(types.get(channel).getObjectValue(connectorSession.getSqlFunctionProperties(), page.getBlock(channel), position));
+                }
+                result.add(unmodifiableList(columns));
+            }
+        }
+        List<List<Object>> results = result.build();
+
+        // Based on com.facebook.presto.server.protocol.Query#getNextResult
+        OptionalLong updateCount = OptionalLong.empty();
+        if (planAndMore.getUpdateType().isPresent() &&
+                types.size() == 1 &&
+                types.get(0).equals(BIGINT) &&
+                results.size() == 1 &&
+                results.get(0).size() == 1) {
+            updateCount = OptionalLong.of(((Number) results.get(0).get(0)).longValue());
+        }
+
+        // successfully finished
+        try {
+            queryCompletedEvent(Optional.empty(), updateCount);
+        }
+        catch (RuntimeException eventFailure) {
+            log.error(eventFailure, "Error publishing query completed event");
+        }
+
+        if (queryDataOutputLocation.isPresent()) {
+            metadataStorage.write(
+                    queryDataOutputLocation.get(),
+                    queryDataJsonCodec.toJsonBytes(new PrestoSparkQueryData(PrestoSparkQueryExecutionFactory.getOutputColumns(planAndMore), results)));
+        }
+
+        return results;
+    }
+
+    public List<Type> getOutputTypes()
```

Comment:
this should only be called at the end and we should throw an exception if fragmentedPlan is null rather than silently ignore